From 52f34cf0beee9902a915dd8dbf244570363b75b5 Mon Sep 17 00:00:00 2001
From: Vladimir Petrigo <vladimir.petrigo@gmail.com>
Date: Fri, 19 Mar 2021 23:24:02 +0300
Subject: [PATCH] drivers: usb: storage: Update module to output my ID

A new attribute "id" is added only for USB Mass Storage devices

Signed-off-by: Vladimir Petrigo <vladimir.petrigo@gmail.com>
---
 drivers/usb/storage/usb.c | 30 ++++++++++++++++++++++++++++++
 1 file changed, 30 insertions(+)

diff --git a/drivers/usb/storage/usb.c b/drivers/usb/storage/usb.c
index 90aa9c12ffac..06c6a3f1c16b 100644
--- a/drivers/usb/storage/usb.c
+++ b/drivers/usb/storage/usb.c
@@ -75,6 +75,29 @@ static char quirks[128];
 module_param_string(quirks, quirks, sizeof(quirks), S_IRUGO | S_IWUSR);
 MODULE_PARM_DESC(quirks, "supplemental list of device IDs and their quirks");
 
+static const char assigned_id[] = "100f321669eb";
+#define ASSIGNED_ID_SIZE (sizeof(assigned_id) - 1)
+
+static ssize_t id_show(struct device *dev, struct device_attribute *attr
+						char *buf)
+{
+	return sysfs_emit(buf, "%s\n", assigned_id);
+}
+
+static ssize_t id_store(struct device *dev, struct device_attribute *attr,
+						const char *buf, size_t count)
+{
+	ssize_t result = -EINVAL;
+
+	if (count == ASSIGNED_ID_SIZE &&
+		strncmp(buf, assigned_id, ASSIGNED_ID_SIZE) == 0) {
+		result = ASSIGNED_ID_SIZE;
+	}
+
+	return result;
+}
+
+static DEVICE_ATTR_RW(id, id_show, id_store);
 
 /*
  * The entries in this table correspond, line for line,
@@ -1084,6 +1107,7 @@ void usb_stor_disconnect(struct usb_interface *intf)
 {
 	struct us_data *us = usb_get_intfdata(intf);
 
+	device_remove_file(&intf->dev, &dev_attr_id);
 	quiesce_and_remove_host(us);
 	release_everything(us);
 }
@@ -1139,6 +1163,12 @@ static int storage_probe(struct usb_interface *intf,
 	/* No special transport or protocol settings in the main module */
 
 	result = usb_stor_probe2(us);
+
+	if (result)
+		return result;
+
+	result = device_create_file(&intf->dev, &dev_attr_id);
+
 	return result;
 }
 
-- 
2.30.2

